### PR TITLE
Refactored cpu manager.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
Rationale for most of the changes is included below, if they're controversial let's discuss them. This PR is just me getting some shapes I had in mind committed to code.

- Renamed "cpuset manager" => "cpu manager" everywhere.
    - Rationale: stated goal in doc of expanding to other controls like cache partitions.
- Introduced a new CPUSet type.
    - Rationale: allow higher-level "set-like" operations to make policy code more readable.
- Factored policy out of cpu manager.
    - Rationale: smaller surface area for extending policies in the future.
- Factored state out of cpu manager policy.
    - Rationale: testability, simplify policy implementations, ease safely exporting state to other packages (or to a persistent checkpoint) in the future. State is read-only outside of the cpu manager package.
- Lifted initial container resources settings and reconciliation loop into the cpu manager.
    - Rationale: reuse
- Normalized log output (but it could use more cleanup.)

cc @sjenning @flyingcougar

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
Tested manually using [cpuset-visualizer](https://github.com/ConnorDoyle/cpuset-visualizer). (Core 0 is unconditionally reserved.)

_Case 1: Guaranteed pod with one container consume 1.0 cpus._

exclusive-1

_Case 2: With one core reserved for a guaranteed container, run a pod in the shared pool._

shared

**Comment by ConnorDoyle** [comment]
cc @derekwaynecarr

File: pkg/kubelet/status/status_manager.go [hunk 1]
```diff
@@ -350,6 +347,7 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 
 // deletePodStatus simply removes the given pod from the status cache.
 func (m *manager) deletePodStatus(uid types.UID) {
+	glog.Infof("[cpumanager] deleting pod status for container (uid: %s)", uid)
```
Comment (sjenning) [inline pkg/kubelet/status/status_manager.go]: cpumanager message outside the cpu manager? seems out of place.

File: pkg/kubelet/status/status_manager.go [hunk 2]
```diff
@@ -163,10 +163,7 @@ func (m *manager) SetPodStatus(pod *v1.Pod, status v1.PodStatus) {
 	m.podStatusesLock.Lock()
 	defer m.podStatusesLock.Unlock()
 	// Make sure we're caching a deep copy.
-	status, err := copyStatus(&status)
-	if err != nil {
-		return
-	}
+	status, _ = copyStatus(&status)
```
Comment (sjenning) [inline pkg/kubelet/status/status_manager.go]: why ignore the err?

**Review by sjenning — COMMENTED** [review]
This is great work!  Thanks a bunch :smile:   I'm inclined to just merge it since this is structured a lot better than my PoC hacky stuff.  Just a couple of questions on the changes in the status manager.

File: pkg/kubelet/status/status_manager.go (hunk 2 above)
Comment (ConnorDoyle) [reply pkg/kubelet/status/status_manager.go]: Not intentional, will revert this part

File: pkg/kubelet/status/status_manager.go (hunk 1 above)
Comment (ConnorDoyle) [reply pkg/kubelet/status/status_manager.go]: Agree, will fix

**Comment by ConnorDoyle** [comment]
Thanks for review @sjenning, OOO today and tomorrow but will get to the updates first thing Wed

**Comment by ConnorDoyle** [comment]
@sjenning updated

**Comment by ConnorDoyle** [comment]
🎉

**Pull request merged**